File: 0210-course-schedule-ii/0210-course-schedule-ii.cpp
class Solution {
public:
    vector<int> findOrder(int numCourses, vector<vector<int>>& prerequisites) {
        unordered_map<int,vector<int>> adjList;
        int length = prerequisites.size();
        
        for(int i=0;i<length;++i){
            int vertex_a = prerequisites[i][0];
            int vertex_b = prerequisites[i][1];
            adjList[vertex_b].push_back(vertex_a);
        }
        
        vector<int> indegree(numCourses);
        for(int course=0;course<numCourses;++course){
            for(int index=0;index<adjList[course].size();++index){
                // cout<<adjList[course][index];
                indegree[adjList[course][index]]++;
            }
        }
        
        queue<int> coursesQueue;
        vector<int> coursesOrder;
        for(int course=0;course<numCourses;++course){
            if(indegree[course] == 0){
                coursesQueue.push(course);
                coursesOrder.push_back(course);
            }
        }
        
        while(!coursesQueue.empty()){
            int currSize = coursesQueue.size();
            while(currSize > 0){
                int currentCourse = coursesQueue.front();
                coursesQueue.pop();
                for(int index=0;index<adjList[currentCourse].size();++index){
                    int neighbourCourse = adjList[currentCourse][index];
                    indegree[neighbourCourse]--;
                    if(indegree[neighbourCourse] == 0){
                        coursesQueue.push(neighbourCourse);
                        coursesOrder.push_back(neighbourCourse);
                    }
                }
                currSize--;
            }
        }
        
        for(int course=0;course<numCourses;++course){
            if(indegree[course] > 0){
                return {};
            }
        }
        
        return coursesOrder;
    }
};

File: 0210-course-schedule-ii/0210-course-schedule-ii.java
class Solution {
    public int[] findOrder(int numCourses, int[][] prerequisites) {
        HashMap<Integer, List<Integer>> graph = new HashMap<>();
        for(int index=0;index<prerequisites.length;++index) {
            int a = prerequisites[index][0];
            int b = prerequisites[index][1];
            graph.putIfAbsent(b, new ArrayList<>());
            List<Integer> curr = graph.get(b);
            curr.add(a);
            graph.put(b, curr);
        }
        int[] indegree = new int[numCourses];
        Queue<Integer> q = new LinkedList<>();
        for(int index=0;index<numCourses;++index) {
            List<Integer> nei = graph.getOrDefault(index, new ArrayList<>());
            for(int i : nei) {
                indegree[i]++;
            }
        }
        for(int index=0;index<numCourses;++index) {
            if(indegree[index] == 0) {
                q.add(index);
            }
        }
        if(q.isEmpty()) {
            return new int[]{};
        }
        int[] order = new int[numCourses];
        boolean[] visited = new boolean[numCourses];
        for(int index=0;index<numCourses;++index) {
            visited[index] = false;
        }
        int currIndex = 0;
        while(!q.isEmpty()) {
            int size = q.size();
            for(int index=0;index<size;++index) {
                int node = q.poll();
                visited[node] = true;
                order[currIndex] = node;
                currIndex++;
                List<Integer> nei = graph.getOrDefault(node, new ArrayList<>());
                System.out.println(nei);
                for(int i : nei) {
                    System.out.println(node + "->" + i);
                    if(visited[i] == true) {
                        return new int[]{};
                    }
                    indegree[i]--;
                    if(indegree[i] == 0) {
                        q.add(i);
                    }
                }
            }
        }
        int count = 0;
        for(int index=0;index<numCourses;++index) {
            if(order[index] == 0) {
                count++;
            }
        }
        if(count > 1) {
            return new int[]{};
        }
        return order;
    }
}


